torch/testing/_internal/common_methods_invocations.py

```diff
@@ -349,6 +349,80 @@ def wrapped_fn(x):
 
     return wrapped_fn
 
+def np_binary_ufunc_type_promotion_wrapper(fn):
+    # Wrapper that passes PyTorch's promoted
+    #   type as an argument to the wrapped NumPy
+    #   binary ufunc when the type promotion
+    #   for the given inputs is inconsistent
+    #   with PyTorch.
+    #   This mimicks PyTorch's type promotion.
+    #
+    # The following describes all scenarios that
+    #   the type promotion is different between
+    #   NumPy and PyTorch
+    # 1) For inputs with integer tensor and float scalar,
+    #   NumPy promotes integer types to double,
+    #   but PyTorch promotes integer types to the
+    #   default scalar type.
+    #
+    # 2) For inputs with two tensors, the table below
+    #   shows the difference
+    #       inputs         NumPy     PyTorch
+    #    int16 float16    float32    float16
+    #    int32 float16    float64    float16
+    #    int64 float16    float64    float16
+    #    int32 float32    float64    float32
+    #    int64 float32    float64    float32
+    #   Special case:
+    #   For two boolean inputs, the result dtype varies
+    #   for different operations in NumPy, but it is
+    #   torch.bool in PyTorch.
+    #
+    # 3) On Windows platform, PyTorch tensor wrapper
+    #   for integer scalar is int64, but NumPy wrapper
+    #   for integer scalar is int32. For boolean tensor
+    #   and integer scalar, NumPy result needs to be
+    #   promoted from int3 to int64
```

Comment:
Repost for the comment:

Facts of binary operations:
1. First argument must be Tensor
2. Second argument can be Tensor or Scalar

There are several cases with different type promotion between NumPy and PyTorch for binary operations:
1. Tensor & Scalar
    - Integer Tensor & Float Scalar
      - np: np.float64
      - torch: default value
    - Boolean Tensor & Integer Scalar on Windows
      - On windows platform, Scalar becomes np.int64 when it isconverted to numpy array. But it becomes torch.int32 when it's converted to Tensor. But, this behavior only impacts the Boolean Tensor, the output types are: 
      - np: np.int64
      - torch: torch.int32
2. Tensor & Tensor
    - Boolean Tensor & Boolean Tensor (varies for different operators)
```python
>>> x = y = np.array(True)
>>> np.fmod(x, y).dtype
dtype('int8')
>>> np.divide(x, y).dtype
dtype('float64')
>>> np.add(x, y).dtype
dtype('bool')
```
```python
>>> x = y = torch.tensor(True)
>>> torch.fmod(x, y).dtype
torch.bool
>>> torch.div(x, y).dtype
torch.float32
>>> torch.add(x, y).dtype
torch.bool
```
    - Integral Tensor & Float Tensor / Float Tensor & Integral Tensor
      - Shown in the table in comment

Reply:
Add Greg's comment here:

> Also, perhaps split out the bool changes since they seem more complicated than the rest?

Reply:
See my comments about disallowing bool above, too, which echo Greg's thinking. Just error on bool inputs in this function, too. I'll make a couple more comments inline.